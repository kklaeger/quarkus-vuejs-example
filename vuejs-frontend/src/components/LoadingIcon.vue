<template>
  <b-loading :is-full-page="false" v-model="isLoading"></b-loading>
</template>

<script lang="ts">
import { Component, Vue } from "vue-property-decorator";
@Component({
  components: {}
})
export default class LoadingIcon extends Vue {
  private isLoading = true;
}
</script>

<style scoped lang="scss"></style>
