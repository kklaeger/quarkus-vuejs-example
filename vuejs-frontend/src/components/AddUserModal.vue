<template>
  <form @submit.prevent="saveUser" class="cardwidth">
    <div class="modal-card" style="width: auto">
      <header class="modal-card-head">
        <p class="modal-card-title">Add User</p>
      </header>
      <section class="modal-card-body">
        <b-field label="Name">
          <b-input type="text" v-model="name" placeholder="Name" required />
        </b-field>
        <div v-if="this.$store.state.addUserIsLoading" class="loading">
          <LoadingIcon />
        </div>
        <div v-if="this.$store.state.addUserHasError" class="error">
          Unable to add user: {{ this.$store.state.addUserErrorMsg }}
        </div>
      </section>
      <footer class="modal-card-foot">
        <button class="button" type="button" @click="$emit('close')">
          Close
        </button>
        <button class="button is-primary" type="submit">
          Save
        </button>
      </footer>
    </div>
  </form>
</template>

<script lang="ts">
import { Component, Vue } from "vue-property-decorator";
import LoadingIcon from "@/components/LoadingIcon.vue";
@Component({
  components: { LoadingIcon }
})
export default class AddUserModal extends Vue {
  private name = "";

  private saveUser(): void {
    this.$store.dispatch("addUser", this.name).then(res => {
      if (res) this.$emit("close");
    });
  }
}
</script>

<style scoped lang="scss">
@import "./../styles/variables.scss";
.cardwidth {
  width: 20rem;
}
@media screen and (max-width: $tablet) {
  .cardwidth {
    width: 100%;
  }
}
.error {
  color: $error-message;
}
</style>
