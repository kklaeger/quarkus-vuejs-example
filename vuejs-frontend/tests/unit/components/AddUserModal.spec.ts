import { Vue } from "vue-property-decorator";
import { mount, createLocalVue } from "@vue/test-utils";
import Vuex from "vuex";
import Buefy from "buefy";
import AddUserModal from "@/components/AddUserModal.vue";
import LoadingIcon from "@/components/LoadingIcon.vue";

const localVue = createLocalVue();
localVue.use(Vuex);
localVue.use(Buefy);

describe(AddUserModal.name + " component", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it("should trigger emit('close') by clicking close button.", async () => {
    const state = {
      users: [],
      fetchUsersIsLoading: false,
      fetchUsersHasError: false,
      addUserIsLoading: false,
      addUserHasError: false,
      deleteUserIsLoading: false,
      deleteUserHasError: false
    };
    const actions = {
      addUser: jest.fn()
    };

    const store = new Vuex.Store({
      state,
      actions
    });

    const wrapper = mount(AddUserModal, { store, localVue });

    wrapper
      .findAll("button")
      .at(0)
      .trigger("click");

    expect(wrapper.emitted().close).toBeTruthy();
    expect(wrapper.find(".error").exists()).toBeFalsy();
    expect(wrapper.find(".loading").exists()).toBeFalsy();
    expect(wrapper.findComponent(LoadingIcon).exists()).toBeFalsy();
    expect(wrapper).toMatchSnapshot();
  });

  it("should update input, click save button and trigger emit('close').", async () => {
    const state = {
      users: [],
      fetchUsersIsLoading: false,
      fetchUsersHasError: false,
      addUserIsLoading: false,
      addUserHasError: false,
      deleteUserIsLoading: false,
      deleteUserHasError: false
    };
    const actions = {
      addUser: jest.fn()
    };

    const store = new Vuex.Store({
      state,
      actions
    });
    (actions.addUser as jest.Mock).mockResolvedValueOnce(true);

    const wrapper = mount(AddUserModal, { store, localVue });

    wrapper.find('input[type="text"]').setValue("test");

    await wrapper
      .findAll("button")
      .at(1)
      .trigger("submit");

    await Vue.nextTick(); // Workaround to fix travis build

    expect(actions.addUser.mock.calls[0][1]).toEqual("test");
    expect(wrapper.emitted().close).toBeTruthy();
    expect(wrapper.find(".error").exists()).toBeFalsy();
    expect(wrapper.find(".loading").exists()).toBeFalsy();
    expect(wrapper.findComponent(LoadingIcon).exists()).toBeFalsy();
    expect(wrapper).toMatchSnapshot();
  });

  it("should update input, click save button and NOT trigger emit('close').", async () => {
    const state = {
      users: [],
      fetchUsersIsLoading: false,
      fetchUsersHasError: false,
      addUserIsLoading: false,
      addUserHasError: false,
      deleteUserIsLoading: false,
      deleteUserHasError: false
    };
    const actions = {
      addUser: jest.fn()
    };

    const store = new Vuex.Store({
      state,
      actions
    });
    (actions.addUser as jest.Mock).mockResolvedValueOnce(false);

    const wrapper = mount(AddUserModal, { store, localVue });

    wrapper.find('input[type="text"]').setValue("test");

    await wrapper
      .findAll("button")
      .at(1)
      .trigger("submit");

    expect(actions.addUser.mock.calls[0][1]).toEqual("test");
    expect(wrapper.emitted().close).toBeFalsy();
    expect(wrapper.find(".error").exists()).toBeFalsy();
    expect(wrapper.find(".loading").exists()).toBeFalsy();
    expect(wrapper.findComponent(LoadingIcon).exists()).toBeFalsy();
    expect(wrapper).toMatchSnapshot();
  });

  it("should display loading icon if addUserIsLoading=true.", async () => {
    const state = {
      users: [],
      fetchUsersIsLoading: false,
      fetchUsersHasError: false,
      addUserIsLoading: true,
      addUserHasError: false,
      deleteUserIsLoading: false,
      deleteUserHasError: false
    };
    const actions = {
      addUser: jest.fn()
    };

    const store = new Vuex.Store({
      state,
      actions
    });
    (actions.addUser as jest.Mock).mockResolvedValueOnce(false);

    const wrapper = mount(AddUserModal, { store, localVue });

    expect(wrapper.emitted().close).toBeFalsy();
    expect(wrapper.find(".error").exists()).toBeFalsy();
    expect(wrapper.find(".loading").exists()).toBeTruthy();
    expect(wrapper.findComponent(LoadingIcon).exists()).toBeTruthy();
    expect(wrapper).toMatchSnapshot();
  });

  it("should display error message if addUserHasError=true.", async () => {
    const state = {
      users: [],
      fetchUsersIsLoading: false,
      fetchUsersHasError: false,
      addUserIsLoading: false,
      addUserHasError: true,
      deleteUserIsLoading: false,
      deleteUserHasError: false,
      addUserErrorMsg: "error"
    };
    const actions = {
      addUser: jest.fn()
    };

    const store = new Vuex.Store({
      state,
      actions
    });
    (actions.addUser as jest.Mock).mockResolvedValueOnce(false);

    const wrapper = mount(AddUserModal, { store, localVue });

    expect(wrapper.emitted().close).toBeFalsy();
    expect(wrapper.find(".error").exists()).toBeTruthy();
    expect(wrapper.find(".error").text()).toEqual("Unable to add user: error");
    expect(wrapper.find(".loading").exists()).toBeFalsy();
    expect(wrapper.findComponent(LoadingIcon).exists()).toBeFalsy();
    expect(wrapper).toMatchSnapshot();
  });
});
