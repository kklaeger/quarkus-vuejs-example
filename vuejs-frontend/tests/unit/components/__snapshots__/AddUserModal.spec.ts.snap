// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddUserModal component should display error message if addUserHasError=true. 1`] = `
<form class="cardwidth">
  <div class="modal-card" style="width: auto;">
    <header class="modal-card-head">
      <p class="modal-card-title">Add User</p>
    </header>
    <section class="modal-card-body">
      <div class="field"><label class="label">Name</label>
        <div class="control is-clearfix"><input type="text" autocomplete="on" placeholder="Name" required="required" class="input">
          <!---->
          <!---->
          <!---->
        </div>
        <!---->
      </div>
      <!---->
      <div class="error">
        Unable to add user: error
      </div>
    </section>
    <footer class="modal-card-foot"><button type="button" class="button">
        Close
      </button> <button type="submit" class="button is-primary">
        Save
      </button></footer>
  </div>
</form>
`;

exports[`AddUserModal component should display loading icon if addUserIsLoading=true. 1`] = `
<form class="cardwidth">
  <div class="modal-card" style="width: auto;">
    <header class="modal-card-head">
      <p class="modal-card-title">Add User</p>
    </header>
    <section class="modal-card-body">
      <div class="field"><label class="label">Name</label>
        <div class="control is-clearfix"><input type="text" autocomplete="on" placeholder="Name" required="required" class="input">
          <!---->
          <!---->
          <!---->
        </div>
        <!---->
      </div>
      <div class="loading">
        <transition-stub name="fade">
          <div class="loading-overlay is-active">
            <div class="loading-background"></div>
            <div class="loading-icon"></div>
          </div>
        </transition-stub>
      </div>
      <!---->
    </section>
    <footer class="modal-card-foot"><button type="button" class="button">
        Close
      </button> <button type="submit" class="button is-primary">
        Save
      </button></footer>
  </div>
</form>
`;

exports[`AddUserModal component should trigger emit('close') by clicking close button. 1`] = `
<form class="cardwidth">
  <div class="modal-card" style="width: auto;">
    <header class="modal-card-head">
      <p class="modal-card-title">Add User</p>
    </header>
    <section class="modal-card-body">
      <div class="field"><label class="label">Name</label>
        <div class="control is-clearfix"><input type="text" autocomplete="on" placeholder="Name" required="required" class="input">
          <!---->
          <!---->
          <!---->
        </div>
        <!---->
      </div>
      <!---->
      <!---->
    </section>
    <footer class="modal-card-foot"><button type="button" class="button">
        Close
      </button> <button type="submit" class="button is-primary">
        Save
      </button></footer>
  </div>
</form>
`;

exports[`AddUserModal component should update input, click save button and NOT trigger emit('close'). 1`] = `
<form class="cardwidth">
  <div class="modal-card" style="width: auto;">
    <header class="modal-card-head">
      <p class="modal-card-title">Add User</p>
    </header>
    <section class="modal-card-body">
      <div class="field"><label class="label">Name</label>
        <div class="control is-clearfix"><input type="text" autocomplete="on" placeholder="Name" required="required" class="input">
          <!---->
          <!---->
          <!---->
        </div>
        <!---->
      </div>
      <!---->
      <!---->
    </section>
    <footer class="modal-card-foot"><button type="button" class="button">
        Close
      </button> <button type="submit" class="button is-primary">
        Save
      </button></footer>
  </div>
</form>
`;

exports[`AddUserModal component should update input, click save button and trigger emit('close'). 1`] = `
<form class="cardwidth">
  <div class="modal-card" style="width: auto;">
    <header class="modal-card-head">
      <p class="modal-card-title">Add User</p>
    </header>
    <section class="modal-card-body">
      <div class="field"><label class="label">Name</label>
        <div class="control is-clearfix"><input type="text" autocomplete="on" placeholder="Name" required="required" class="input">
          <!---->
          <!---->
          <!---->
        </div>
        <!---->
      </div>
      <!---->
      <!---->
    </section>
    <footer class="modal-card-foot"><button type="button" class="button">
        Close
      </button> <button type="submit" class="button is-primary">
        Save
      </button></footer>
  </div>
</form>
`;
